### ISSUE TITLE
Multi-task vs. multi-fidelity: which should I choose?

### ISSUE THREAD

**Issue by sgbaird** [facebook/Ax#1038]
Both seem to follow a similar setup.

[Multi-task](https://ax.dev/tutorials/multi_task.html) (Ax):

> A typical use case is optimizing an expensive-to-evaluate (online) system with supporting (offline) simulations of that system.

[Multi-fidelity](https://botorch.org/tutorials/discrete_multi_fidelity_bo) (BoTorch):

> In this tutorial, we show how to do multi-fidelity BO with discrete fidelities based on [1], where each fidelity is a different "information source." This tutorial uses the same setup as the [continuous multi-fidelity BO tutorial](https://botorch.org/tutorials/multi_fidelity_bo), except with discrete fidelity parameters that are interpreted as multiple information sources.
> 
> We use a GP model with a single task that models the design and fidelity parameters jointly. In some cases, where there is not a natural ordering in the fidelity space, it may be more appropriate to use a multi-task model (with, say, an ICM kernel). We will provide a tutorial once this functionality is in place.

Going to close this outright as I did find an answer to the question in the BoTorch docs, namely that multi-fidelity is probably better than multi-task when there is a natural ordering in the fidelity space. Maybe worth mentioning that I'm guessing constraints are supported for multi-task. In contrast, constraints are not yet implemented for multi-fidelity #961 and discrete fidelities are not fully exposed in Ax (but are present in BoTorch) #979. Not sure if anyone has tried comparing the performance of multi-fidelity vs. multi-task for a fidelity problem with "a natural ordering".

**Comment by Balandat** [comment]
You're right that these are very related and sometimes used in similar ways. In my mind the main distinction is what exactly one is optimizing. 

A multi-task approach essentially uses data from multiple different related (hopefully well correlated) outcomes to improve the model over what would be possible with using only the data from the "target task", i.e. the task of interest. In such a setting you may or may not be able to choose evaluating tasks other that the target task (maybe this is just previously generated data that you have sitting around but can't get more of). 

In a multi-fidelity setting, you do have that authority, i.e. you get to choose at which fidelity you want to make observations at. Typically the goal is optimizing performance at the highest fidelity, so the name of the game then becomes not just what configuration to evaluate but also at which fidelity (and how that helps you learn about the maximum fidelity outcome that you care about). 

So in a nutshell you could say that every multi-fidelity setting is also a multi-task setting, but the converse is not true. 

Also, multi-task often suggests as you indicated that the tasks are discrete or categorical without natural ordering, though that's not necessarily always the case.

**Comment by sgbaird** [comment]
@Balandat this helps. I like the clarification regarding the authority to evaluate certain task types depending on the setting. Thank you!

**Comment by sgbaird** [comment]
> In [a multi-task] setting you may or may not be able to choose evaluating tasks other that the target task (maybe this is just previously generated data that you have sitting around but can't get more of).

Something that I may have overlooked upon first reading was the "may or may not" from the above snippet, but I think I get the idea:
- multi-fidelity treats the fidelity parameter (and by extension, cost) as part of the model, so it's suitable for tasks that have a natural ordering to them
- multi-task treats measurements are treated as correlated, but in an unordered sense, so one task is not necessarily "better" than another, and the cost (CPU time, $$$, etc.) isn't directly treated by the model at all

It makes sense that tasks with a continuous, natural ordering (i.e., fidelity parameter) would be more likely to be ones where the user gets to choose which task to run. It also makes sense that tasks that are correlated and unordered are often ones where the user may not be able to choose to run other tasks. For example, "data that you have sitting around" or measurements from separate databases.

The multi-task notebook uses a "biased but highly correlated" function for the offline measurements:

<img src=https://github.com/facebook/Ax/assets/45469701/b9f72323-c3fa-4182-9aeb-c6bd35947efd width=300>

I take it that multi-fidelity problems could also have tasks that are "biased but highly correlated", and the model would naturally learn about the bias and correlations through the modeling of the fidelity parameter.

On a related note, what does "offline" vs. "online" mean? In the context of this discussion, I could see offline meaning that new measurements can't be made, whereas online means that new measurements can be added.

**Comment by sgbaird** [comment]
For my own sake, use cases in the physical sciences that, in my mind, clearly fall into the multi-task vs. multi-fidelity are:

multi-fidelity:
- Case 1: Timestep for a molecular dynamics simulation. Longer timesteps lead to faster, less accurate simulations. Shorter timesteps lead to slower, more accurate simulations. The computational scientist varies temperature, pressure, and catalysts to maximize the rate at which some chemical reaction occurs.

multi-task:
- Case 2: Using existing Materials Project data for properties correlated with superhardness (e.g., bulk modulus and shear modulus). The experimentalist isn't able to run new simulations, so the simulation data is considered fixed. The experimentalist uses only the simulations with the most stable crystal structure. After filtering, only the chemical formula (composition) is used as a model input. The experimentalist varies composition while keeping processing conditions fixed and attempts to maximize the experimentally measured hardness value.

The following case at first seems like multi-task, but maybe actually fits into contextual optimization:

- Case 3: Same as Case 2, except the experimentalist wants to **vary processing conditions** such as temperature and anneal time. Importantly, **the simulations do not consider processing information**. So, the input features are somewhat different between the simulations and the wetlab experiments. This is where contextual optimization comes in. The experimenter uses the ~10k simulation datapoints to train a model to predict bulk and shear moduli and then uses the predictions as contextual features during the optimization. Two routes for this are:
  - Generate a list of predefined candidates, predict the bulk and shear moduli for each candidate, then compute the acquisition function value for each candidate https://github.com/facebook/Ax/issues/905 https://github.com/facebook/Ax/issues/771
  - Something fancier that still allows for continuous optimization of the acquisition function (some kind of stacked model? predict these features at all MC points and wherever else needed internally?). AFAIK not directly supported in Ax, and maybe not even straightforward in BoTorch

**Comment by Balandat** [comment]
> On a related note, what does "offline" vs. "online" mean? In the context of this discussion, I could see offline meaning that new measurements can't be made, whereas online means that new measurements can be added.

So in this discussion "offline" and "online" are meant quite literally. We often run online A/B tests to evaluate potential changes to infrastructure of products, and these are relatively "costly" in the sense that they requiring the tests on actual traffic. But we may also have the possibility of using some offline simulator that can simulate the effect of the proposed changes. It's much cheaper to run the simulator, so that's the "low fidelity, low cost" observation.
